### ISSUE TITLE
Autofix refactors

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#7784]
Requests:
- https://github.com/stylelint/stylelint/issues/2643
- https://github.com/stylelint/stylelint/issues/7192
- https://github.com/stylelint/stylelint/issues/4664

Supporting tasks:
- https://github.com/stylelint/stylelint/issues/7787
- https://github.com/stylelint/stylelint/issues/7788
- https://github.com/stylelint/stylelint/issues/7836

(@Mouvedia were there others you think are related to this effort?)

---------

Thanks to the efforts by @Mouvedia we have already made a lot of progress on #2643 and I think that in doing so we also have a good active mental model for the parts involved.

I think that it would be good at this time to take a moment to evaluate where we want to end up.
I also want to ensure that we are happy with both the public interface and the internal implementations.

---------

###  Add stylelint-disable commands support to autofix #2643

- fixes should be gated behind range checks
- range checks must align with disable range checks for warnings

### Add replaced text and corresponding range to `Warning` type #7192

- no fixes can be applied
- data to enable an external driver to apply fixes must be exposed
- interop with other tools makes this feature more useful

### Add fixed field to Result object #4664

- for each fix we must also capture a report so that users can see why something was fixed

### Fix data as part of warnings

ESLint includes data for unfixed, but fixable issues in lint messages.
This approach makes it easy to resolve #7192.

But at the same time it makes it impossible to resolve #4664.

We can't expose fixed issues as part of warnings because the source of the warning would already have been mutated. Doing so anyway implies that all surrounding tools (editor integrations, formatters, ...) would also need to update.

At this time I would challenge the need to resolve #4664.
I think it is more important to solve the other requests and to align with tools like ESLint.

----------

#### exposed as: 
- https://eslint.org/docs/latest/integrate/nodejs-api#-editinfo-type

This is a `fix` property on `LintMessage` with type [`fix (EditInfo | undefined)`](https://eslint.org/docs/latest/integrate/nodejs-api#-editinfo-type)

#### implemented as:
- https://eslint.org/docs/latest/extend/custom-rules#applying-fixes

```js
context.report({
    node: node,
    message: "Missing semicolon",
    fix(fixer) {
        return fixer.insertTextAfter(node, ";");
    }
});
```

#### Note worthy:
- data about fixes is part of lint messages, it is not something separate
- `fix()` provides a `fixer` object with helper methods for mutations
- anything in `fix()` doesn't actually mutate the original source. Only the edits to be made are captured. See : https://github.com/eslint/eslint/blob/44915bb95dfa21f946021d77b3b361e7e9b140e0/lib/linter/rule-fixer.js#L24-L29

---------

### current proposed API for `fix` as part of warnings:

see : https://github.com/stylelint/stylelint/issues/7788#issuecomment-2186850056

```ts
type EditInfo = {
	replacement: string;
	range: Range;
};

type Warning = {
	fix?: EditInfo;
};
````

**Comment by romainmenke** [comment]
At first glance I think we have a couple of issues with the [current](https://github.com/stylelint/stylelint/tree/c04b77d9781100421634157e780665722980519f) state.

- #7192 and #4664 require that data about fixes is exposed as part of the warnings
- #7192 requires that data about fixes is exposed without mutating the original source or the AST

**Comment by Mouvedia** [comment]
> no fixes can be applied

~~By that you mean, the data is available without having to fix?
i.e. a dry run~~

You have answered on the other issue that:

> We must acquire the replacement text without mutating the AST.

That would mean that it would be the responsibility of the warn path.
i.e. not the fix path

Said another way it will require to pass a new property to `report`.
Are we on the same page?

**Comment by romainmenke** [comment]
At this time I first want to ensure that we have a good understanding of what we are fixing.

Discussing how to address potential issues can be done later.
There are always many ways to tackle features/bugs, but conversations become messy when we are discussing **what**, **why** and **how** at the same time :)

**Comment by romainmenke** [comment]
In other words

- are there goals/constraints that I have overlooked?
- are there goals/constraints that someone does not agree with?
- are there other open issues that might influence this effort?

**Comment by Mouvedia** [comment]
For the constraint "no fixes can be applied" I wanted to clarify that you realise that it implies that it will require to pass a new property to `report`.
i.e. it's not only added if you are fixing

If that's obvious to you, apart from that I would like to know what you mean by

> fixes should be gated behind range checks

Do you mean that we require `line` or `node` to be passed to `report`? (which is already the case)

**Comment by romainmenke** [comment]
> For the constraint "no fixes can be applied" I wanted to clarify that you realise that it implies that it will require to pass a new property to `report`.
i.e. the data is not only added if you are fixing

I agree that this implies further changes.

However, can already see multiple ways to achieve this goal and each has different implications for code complexity, performance, API complexity, ... Not all of these options would introduce a new property that need to be passed to `report()`

--------

> fixes should be gated behind range checks

This has already been implemented (by you :) )
It means that fixes should only be applied when they aren't inside the ranges of disable comments.

**Comment by ybiquitous** [comment]
@romainmenke Thank you so much for arranging the issues!

In my honest opinion, we should focus on #2643 refactoring first. In the refactoring, we can leave that `fix()` doesn't return anything. I think it's easier just to replace `context.fix` with `fix()` than calculating the range.

After the refactoring finishes, I'd like to release a new version to avoid a big-ban release.

**Comment by romainmenke** [comment]
My concern is that we can't ship at this time because `fix()` doesn't yet have the needed shape.

At this time I think it should look something like this:

```ts
fix?: (applyFixes: boolean) => Array<{range: Range, text: string}>;
```

I want to avoid shipping an update that implies refactors for plugin authors, only to ship another one that implies more refactors a few weeks later.

-------

> After the refactoring finishes, I'd like to release a new version to avoid a big-ban release.

Can we go depth first instead of breadth?

We do not need to refactor all rules (breadth) before a release.
But I do think we should try to ship a stable and future proof version of `report()` and `fix()` (depth)

**Comment by Mouvedia** [comment]
> In the refactoring, we can leave that fix() doesn't return anything.

The first 2 refactors were done that way. The return type of the fixer permits the plugin authors to return `void` to facilitate/smooth the transition.

> I think it's easier…

It would be indeed. If that motivates you or @romainmenke to take part in the effort we could relax that requirement for PRs.

> After the refactoring finishes, I'd like to release a new version to avoid a big-ban release.

At this pace, without some help, it will take 2 to 3 releases.

> We do not need to refactor all rules (breadth) before a release.

That was already discussed with @ybiquitous and he was OK if it was to be spread between several releases.

**Comment by romainmenke** [comment]
> I think it's easier just to replace `context.fix` with `fix()` than calculating the range.

🤔 My main concerns:
- causing too much churn for plugin/rule authors with consecutive changes to `fix()` / `report()`
- being locked in because 3rd parties are already using it

Can we keep this as an internal feature for now?
If 3rd parties aren't using this (yet), then we have more freedom to make changes in the future.

**Comment by Mouvedia** [comment]
@romainmenke 

> `=> Array<{range: Range, text: string}>;`

This implies that we don't store the new range.
Are you fine with not being able to test the ranges—like we are doing for warnings currently—of the `text`.
<sub>I suppose that `text` is the new text in your example.</sub>

> `=> Array<…>;`

Can you provide at least one rule that would benefit from being just an `Array`?
If not, we should be more careful and opt for `{range: Range, text: string} | Array<{range: Range, text: string}>`.

**Comment by ybiquitous** [comment]
@romainmenke I agree. We should not publish `fix()` yet. This `fix()` signature may change since we're still discussing it.

> Can we keep this as an internal feature for now?

Yes, we can. 👍🏼 I suggest marking the `Problem#fix` method as *internal* like this:

```ts
/**
 * @internal
 * WARNING: Don't use this feature now. It may change in the future.
 */
fix?: () => void | never | Range;
```

https://github.com/stylelint/stylelint/blob/2da2c22b435fa78c2a5a2c745e5c0dcd22f0fd4a/types/stylelint/index.d.ts#L633

At the same time, we don't write any documents about `fix()` yet.

**Comment by Mouvedia** [comment]
> we don't write any documents about `fix()` yet

I am expecting that the return type of `fix` will be set in stone by the time we do the last refactor.
As long as `| void` remains I am good with anything sensible.
i.e. we should not demand to all plugin authors to return ranges/text
The documentation should give examples that return fix data and link to one of our rule.

</hr>

Once @romainmenke answers my questions in [this comment](https://github.com/stylelint/stylelint/issues/7784#issuecomment-2183235190), we should be on the page concerning goals/constraints.

**Comment by romainmenke** [comment]
> This implies that we don't store the new range.
Are you fine with not being able to test the ranges—like we are doing for warnings currently—of the text?

I still don't know why we would need the new range and no one has answered on **why** when I asked for this :)

I am fine with and prefer not storing the new range.

-------

> Can you provide at least one rule that would benefit from being just an `Array`?
If it's only one, we should opt for `{range: Range, text: string} | Array<{range: Range, text: string}>`.

That is an implementation detail and as I said, I am likely wrong about the details in the example :)

The purpose of that example was purely to illustrate that I expect breaking changes to the signature of this function. The example is based on my current view on the requested features, but it is limited by what we currently implemented and tested. Further thinking and developments will likely change my views on this.

**Comment by Mouvedia** [comment]
> > This implies that we don't store the new range.
> > Are you fine with not being able to test the ranges—like we are doing for warnings currently—of the text?
> 
> I still don't know why we would need the new range and no one has answered on **why** when I asked for this :)
> 
> I am fine with and prefer not storing the new range.

That's not what I have asked. I know you are fine with not storing the new range.
Once we don't, I am saying that we won't be able to test the expected range in our test suite.

e.g.
```
	code: 'a {}',
	fixed: 'foo {}',
	message: '',
	line: 1,
	column: 1,
	endLine: 1,
	endColumn: 2,
      bar: {
		line: 1,
		column: 1,
		endLine: 1,
		endColumn: 4,
      }
```

If you are fine with that aspect, we can move on.

**Comment by romainmenke** [comment]
> Once we don't, I am saying that we won't be able to test the expected range in our test suite.

That is correct, no part should return, record or expose the new ranges.
So we also do not need test coverage for these.

(please do tell me if I am wrong here and if these new ranges are needed for something I am overlooking)

But this also seems like an implementation detail and not a goal/constraint?
This is not something that will drive the design, right?

**Comment by Mouvedia** [comment]
> But this also seems like an implementation detail and not a goal/constraint?

It's not a detail because once the return type of `fix` is established, it will require plugin authors another refactor to support whatever _new range_ would enable. I just gave you an internal usage as an example.
If you are confident about this, yes we could return `text` directly.
Since you said that

> We must acquire the replacement text without mutating the AST.
> …
> I can already see multiple ways to achieve this goal […] Not all of these options would introduce a new property that need to be passed to report()

I am really curious about these options because it will have an impact on the shape of the object returned by the fixer.
e.g. `{ mutate: () => void, text: string, … }`
If you are not passing it to `report` what are the alternatives?

**Comment by romainmenke** [comment]
If `fix()` is considered private then this is all just an internal implementation detail.
We can change the return type as much as we want up until we make it part of the public API.

**Comment by Mouvedia** [comment]
In #7786, I have laid out a possible solution that should cover all the goals/constraints.

It is either

1.

```ts
	type FixerData = {
		range: Range | Range[];
		fixed: boolean;
		text?: string | string[];
	};

	interface Fix {
		(): void | never;
		editInfo?: {
			range?: Range;
			text: string | string[];
		};
	}
```

2.

```ts
	type FixerData = {
		range: Range | Range[];
		fixed: boolean;
		text?: string | string[];
	};

	export type Problem = {
		/* … */
		fix?: () => void | never;
		editInfo?: {
			range?: Range;
			text: string | string[];
		};
	};
```

I have chosen 1. in the PR but 2. would be acceptable as well.
I have rejected `fix?: () => { mutate: () => void | never, editInfo: EditInfo };` because it would too complicated for plugin authors IMHO.

Can I have your feedback?
Is it acceptable?
What should be the name? (`editInfo` currently)

**Comment by romainmenke** [comment]
[I've changed my mind about the array stuff :)](https://github.com/stylelint/stylelint/issues/7788)

You can change `3` to :

```ts
export type Problem = {
	/* … */
	fix?: (applyFixes: boolean) => { range: Range; text: string; }
}
```

**Comment by romainmenke** [comment]
I've updated the original post with references from ESLint, more sub tasks and conflicts between the different requests.

-------

Some more digging into the observable behavior of ESLint :

<details>
<summary>ESLint `verify` vs. `verifyAndFix` for fixable and unfixable issues</summary>

```js
const { Linter } = require('eslint');

{ 
	const linter = new Linter();

	const code = 'var b = !!foo; function bar() { var b = 1; }';
	const messages = linter.verify(
		code,
		{
			rules: {
				'no-implicit-coercion': 'error',
				'no-shadow': 'error',
		} },
		{ filename: 'foo.js' },
	);

	console.log(code);
	console.log('----------------');
	console.log('');
	console.dir(messages, { depth: null });
	console.log('----------------');
}

{
	const linter = new Linter();

	const code = 'var b = !!foo; function bar() { var b = 1; }';
	const messages = linter.verifyAndFix(
		code,
		{
			rules: {
				'no-implicit-coercion': 'error',
				'no-shadow': 'error',
			}
		},
		{ filename: 'foo.js' },
	);

	console.log(code);
	console.log('----------------');
	console.log('');
	console.dir(messages, { depth: null });
	console.log('----------------');
}

```
</details>

**Comment by ybiquitous** [comment]
Based on `EditInfo` decided in #7788, I prefer the option `3`. For example:

```ts
type EditInfo = {
	replacement: string;
	range: Range;
};

type Problem = {
	fix?: (applyFixes: boolean) => EditInfo | void;
};
```

But, what does the `applyFixes: boolean` argument mean?

**Comment by Mouvedia** [comment]
> what does the applyFixes: boolean argument mean?

This is important.
Ill show you by changing the POC PR.
i.e. it's passed to the `fix` function inside `isFixApplied`

**Comment by romainmenke** [comment]
Yes, this is the aspect I am most conflicted about myself aswel.

The request in https://github.com/stylelint/stylelint/issues/7192 is to capture and expose all the data for fixes but without actually applying any changes.

ESLint has a nicer solution for this because they expose a helper to make the mutations.

see : https://eslint.org/docs/latest/extend/custom-rules#applying-fixes

```js
context.report({
    node: node,
    message: "Missing semicolon",
    fix(fixer) {
        return fixer.insertTextAfter(node, ";");
    }
});
```

Depending on config they apply the fixes or not, they don't leave it up to rule authors to get this right.

---------

I introduced the concept of `applyFixes` because I thought this could reduce duplication in the code for rules.

The more time I spend thinking about this, the more complexity I discover.

Making changes to an AST and serializing the whole AST to apply a fix is relative simple.

Determining the entire replacement text and the ranges without changing the AST is very tricky. (see : https://github.com/stylelint/stylelint/pull/7802#discussion_r1653241078)

Requirements for warnings that include fix data:
- no changes to the source code
- no changes to shared AST's (although cloning and then mutating is fine)
- ranges are crucial
- each fix must correspond with exactly one warning

Requirements for fixes:
- source code is fixed
- changes to AST's are fine and even needed
- ranges are irrelevant
- no warnings are produced for applied fixes

------

Finding the right way forwards is not trivial.

**Comment by Mouvedia** [comment]
> If plugin author forgets to handle `applyFixes`, `fix()` would be unexpectedly called, right?

@ybiquitous For this, you have 2 more solutions apart from passing a `helper` like ESLint.
Both are not to my liking though.

First one is having the return type of `fix` changed to
```
{
  range?: Range;
  text:  string;
  mutate: () => void | never; 
}
```
which would result
```js
report({
  //…
  fix: () => ({
    mutate: () { some_node.value = fixed; },
    text: 'bar',	
  })
}
```
for your example.

The second is switching the default of the boolean.
i.e. renaming it `doNotApply` so that when `true` is _not_ passed the fix is applied by default

**Comment by ybiquitous** [comment]
Sorry for my late response. I have been considering what a better API is.

To maximize usability for plugin authors, I still want to stick to an ESLint-like API. Please take a look at the example below.

We have used directly the [PostCSS API](https://postcss.org/api/) to mutate a AST node like this. But, instead of doing this:

https://github.com/stylelint/stylelint/blob/cb3a7f8d5e89f2e467a4166293c17d3c2d1e72cc/lib/rules/alpha-value-notation/index.mjs#L102-L105

Can we introduce a new mutation API like this?

```js
report({
  fix(fixer) {
    alpha.value = fixed;
    return fixer.setDeclarationValue(decl, parsedValue.toString()); 
  }
});
```

**Comment by Mouvedia** [comment]
> Can we introduce a new mutation API like this?
> …
> `return fixer.setDeclarationValue(decl, parsedValue.toString());`

This would be impractical or would require a lot of work to support all mutation permutations currently available to all rules.
i.e. `fixer` would have so many methods implemented that it would be ridiculous

A pragmatic alternative could be the proposal that I discarded initially: the `fix` function would have a mutation callback in the object returned.
e.g.
```js
report({
  // …
  fix: () => ({
	mutate: () => {
        alpha.value = fixed;
	    setDeclarationValue(decl, parsedValue.toString()); 
    },
    replacement: 'bar',
    // optional
    //range,
  })
});
```

**Comment by ybiquitous** [comment]
But in your proposal, plugin authors would have to specify `replacement` in addition to `parsedValue.toString()`, right?

**Comment by Mouvedia** [comment]
In all proposals that have to cover #7192 requirements, you have to provide the new text slice and an accurate `range` _if_ the warning range doesn't already match it.

i.e. any proposal need at least

- an optional range
- a string
- a function which runs the mutation code

Romain ditched the multi-range feature, so that's what the 4 proposals aims to do.

summary of the proposals:

1. pass to report (add properties to `Problem`)
2. add to `fix` callback (like we are doing for rules)
3. `fix` returns an object but has a boolean argument
4. `fix` returns an object containing a mutation callback

**Comment by Mouvedia** [comment]
> But in your proposal, plugin authors would have to specify replacement in addition to parsedValue.toString(), right?

> I'm asking about the proposed `fix()` API example in [#7784 (comment)](https://github.com/stylelint/stylelint/issues/7784#issuecomment-2206272847)

And I am saying that all proposals should pass a string—alongside the `range` if required—so that it can be exploited by tools. So yes, whatever the name—e.g. `text` or `replacement`—it must be provided.
Is the requirement of passing that string a problem? Is it too strict in your opinion?

**Comment by romainmenke** [comment]
I also keep coming back to something like @ybiquitous proposes in : https://github.com/stylelint/stylelint/issues/7784#issuecomment-2206210487

Providing specific helpers makes it possible to abstract away a lot of complexity.
Rule authors would only need to provide a part of the information and the Stylelint library would infer the rest.

I am curious what the challenges would be for this.

A big difference between ESLint and Stylelint is that we rely on other parsers for parts of the AST. value parser, media query parser, ...

This multi-tiered approach to parsing and mutating is difficult to provide helpers for.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for the help. I feel ESLint's API is easy to use, but the currently proposed `fix()` API is not.

I also want to know the difficulties if we adopt an ESLint-like API.

Backing to my question, I'm not against passing a string. I'm just wondering if the following API might not be easy to use:

```js
report({
  fix() {
    const replacement = parsedValue.toString();

    return {
      mutate() {
        alpha.value = fixed;
        setDeclarationValue(decl, replacement); 
      },
      replacement,
      range: // calculating range...
    };
  }
});

**Comment by Mouvedia** [comment]
> I am curious what the challenges would be for this.

For the majority of the fixable core rules, it would be easy.
The rest though, I am not sure.
The problems lies in the fact that we never restricted what can be done regarding mutations until now.
i.e. we don't know what the plugin rules are doing

We can try to cover all possible cases but it would be safer to have an anything goes policy.
Or at the very least an escape hatch for the cases that we didn't think of if we were to provide an helper, we would need a callback that let's you do what you want.

tl;dr: I am saying that an helper would be impractical if it's not complete (and it probably cannot be)

**Comment by romainmenke** [comment]
The complexity and issues we are current discussing does make me wonder if we should instead consider switching to ESLint 😅 

- https://eslint.org/blog/2024/07/whats-coming-next-for-eslint/#language-plugins
- https://github.com/stylelint/stylelint/issues/6593

I am really happy with the `fix` function as a way to have working disable comments.
But the remaining tasks feel really complex and impactful (potentially breaking) to me.

It might be interesting to defer these changes until ESLint has progressed more?
I am curious how easy or hard it would be to port Stylelint rules to ESLint.

**Comment by Mouvedia** [comment]
I think the most non-breaking would be proposal 1.
It's very straightforward and kinda makes sense.
We are adding new properties to `Problem`.

i.e.
```
report({
  // …
  fix,
  replacement,
  range, // optional
});

// or

report({
  // …
  fix,
  foo: { // name to bikeshed
    replacement,
    range, // optional
  }
});
```

Nothing fancy or complicated.
i.e. we let consumers do whatever they want in the `fix` callback

**Comment by ybiquitous** [comment]
If ESLint became an infra, Stylelint could focus on implementing rules. I hope this will be realized in the future, but I assume there are still challenges. Let's keep an eye on the ESLint team's work.

**Comment by ybiquitous** [comment]
> I think the most non-breaking would be proposal 1.
> It's very straightforward and kinda makes sense.
> We are adding new properties to `Problem`.

As #7836 describes (thanks @romainmenke 😉), this is not as straightforward as we thought. There are some challenges like performance, usability for plugin authors, etc.

**Comment by romainmenke** [comment]
Closing as we have unblocked progress and have been following up in individual issues.
Thank you everyone for your thoughts and input!
